django_extensions/db/fields/encrypted.py

```diff
@@ -0,0 +1,101 @@
+# -*- coding: utf-8 -*-
+import base64
+import warnings
+
+import django
+from django import forms
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured
+from django.db import models
+
+try:
+    from cryptography.fernet import Fernet
+except ImportError:
+    raise ImportError('Using an encrypted field requires the cryptography module. '
+                      'You can obtain cryptography from https://cryptography.io/.')
+
+
+class EncryptionWarning(RuntimeWarning):
+    pass
+
+
+class BaseEncryptedField(models.Field):
+
+    def __init__(self, *args, **kwargs):
+        if not getattr(settings, 'DJANGO_EXTENSIONS_ENCRYPTION_FIELDS_KEY', None):
+            raise ImproperlyConfigured('You must set the settings.DJANGO_EXTENSIONS_ENCRYPTION_FIELDS_KEY.')
+
+        self.key = getattr(settings, 'DJANGO_EXTENSIONS_ENCRYPTION_FIELDS_KEY')
+        self.fernet = Fernet(self.key)
+
+        # Encrypted size is larger than unencrypted
+        self.unencrypted_length = max_length = kwargs.get('max_length', None)
+        if max_length:
+            kwargs['max_length'] = self.calculate_crypt_max_length(max_length)
+
+        super().__init__(*args, **kwargs)
+
+    def calculate_crypt_max_length(self, unencrypted_length):
+        # TODO: Re-examine if this logic will actually make a large-enough
+        # max-length for unicode strings that have non-ascii characters in them.
+        # For PostGreSQL we might as well always use textfield since there is little
+        # difference (except for length checking) between varchar and text in PG.
+        test_bytes = ('x' * unencrypted_length).encode()
+        encrypted_data = base64.urlsafe_b64encode(self.fernet.encrypt(test_bytes)).decode()
+        return len(encrypted_data)
+
+    def to_python(self, value):
+        if value:
+            decrypted_text = self.fernet.decrypt(
+                base64.urlsafe_b64decode(value)
+            )
+            retval = decrypted_text.decode()
+        else:
+            retval = value
+        return retval
+
+    def from_db_value(self, value, expression, connection):
+        return self.to_python(value)
+
+    def get_db_prep_value(self, value, connection, prepared=False):
+        # Truncated encrypted content is unreadable,
+        # so truncate before encryption
+        max_length = self.unencrypted_length
+        if max_length and len(value) > max_length:
+            warnings.warn("Truncating field %s from %d to %d bytes" % (
+                self.name, len(value), max_length), EncryptionWarning
+            )
+            value = value[:max_length]
+
+        encrypted_data = self.fernet.encrypt(value.encode())
```

Comment:
Should this be relying on UTF-8 encoding? and also storing the data as base64 encoded binary? I'm not familiar with keyczar, but I _think_ this should probably be using a `BinaryField` and not base64 encoding, for the input if this is a `EncryptedCharField` you may want to have the user be allowed to pass the encoding (so this base class would have a `self.encoding` which could default to UTF-8), but I would think you might want to also provide a `EncryptedBinaryField` which has `self.encoding` of `None` which would have this base class not try to decode the data and just return the bytes directly.

Reply:
Good questions... my first though is to have multiple variants for this...

I would indeed want the data stored in `BinaryField` most of the time.... (since that makes a lot of sense to me and cuts down on overhead)...

When I do want so much portability that I see the need for using some kind of character field I would actually think 99/100 for me the size (max_length) does not matter.. Since I almost exclusively use PostgreSQL there is little difference between `varchar` and `text` for me.

However I can understand that there are also use-cases for people out there wanting  a `CharField(max_length=123)`...

On another note... would it also not make sense to start storing these things like we do with hashed passwords ?

Much like `User.password` uses `pbkdf2_sha256$180000$fgB0GTcbB2Rg$rpkfa5oqz4RHqZYS+Bsm3ZPNSoVUWU3mWitXI1fxuto=`

Might we not want to store something like:

`fernet$1$gAAAAABfIdWwisKpNayqvyyrK0TmsUjf2vZWTgpEupiGipi2t2V-KZFLXsekO5Bfh2TaNCeyFtKW6_NwP1u0dEH3XhlpQ9rAYA==`

Where first value is encryption type `fernet`, the second is flag / version field for forward compatibility and the third value is the encrypted string itself (here base64 encoded).